### PR TITLE
Final content for the first blog 

### PR DESCRIPTION
**Pull request by hackintoshrao**
- Blog on getting started with Apache Kafka.
- The first part deals with conveying the need for a system like Kafka and how does it fit along with rest of the pieces in the backend.
- Second part deals with using installing and using Apache Kafka.

### CODE REVIEW

**Comment by zsherman** [comment]
thanks @hackintoshrao! I'm going to have @sidhanthp take a look through and make any comments as necessary 👍

File: posts/hello-world-in-kafka-using-python.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+This might look simple but your solution is not complete. There are multiple challenges when trying to solve this problem. Let us try to understand some of them:
+
+1. Events like clicks, requests, impressions and searches results in high frequency interaction/requests or data flow to your webserver and your primary database which is meant to serve your product features may be equipped to store high frequency writes or scale seamlessly and this would introduce high latencies as more and more events pour in to the server.
+2. If you still go ahead and store high frequency in database systems like SQL or Mongo it would be really hard to introduce and reconstruct a new system or a database on all of the historical data.You pretty much lose the flexibility to extend the capabilities of your system by introducing new techologies.
+3. What if you have data processing systems in place to process these events to gain deeper insights? Since these systems wouldn't be capable of handling high frequency reads and you wouldn't have access to the true source of data it is practically impossible to experiment with various data processing, machine learning algorithms on all of the data.
```
Comment (sidhanthp) [inline posts/hello-world-in-kafka-using-python.md]: What do you mean when you say "you wouldn't have access to the _true source of data_"? I see you wrote below that we have 'all the different systems residing in one single place', can't you still have a single source of data by joining together sql tables?

File: posts/hello-world-in-kafka-using-python.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+3. Each of the above applications, whether they are producing the data or consuming the data, might follow their own data formats which means that you will need additional systems for data transformations when there is exchange of data across these applications and your database.
+
+
+All these problems can be better addressed by bringing a streaming platform like Kafka into the picture. A streaming platform is a system that can perform the following:
+
+1. Store huge amount of data that can be persistent, checksummed and replicated for fault tolerance
```
Comment (sidhanthp) [inline posts/hello-world-in-kafka-using-python.md]: Explain why we want our data to be 'persistent, checksummed, and replicated'

File: posts/hello-world-in-kafka-using-python.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+1. Download the latest 1.1.0 release of [Kafka](https://www.apache.org/dyn/closer.cgi?path=/kafka/1.1.0/kafka_2.11-1.1.0.tgz)
+2. Un-tar the download using the following command:
+`tar -xzf kafka_2.11-1.1.0.tgz`
+3. cd to kafka directory to start working with it:
+`cd kafka_2.11-1.1.0`
+
+## Starting the Server
+
+Kafka makes use something called as ZooKeeper which is a centralized service for a distributed environment like Kafka. It offers configuration service, synchronisation service and naming registry for large distributed systems. You can read more about it [here](https://zookeeper.apache.org/).
```
Comment (sidhanthp) [inline posts/hello-world-in-kafka-using-python.md]: Why is ZooKeeper necessary and what does it do in layman terms? Don't just link them to the documentation.

File: posts/hello-world-in-kafka-using-python.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+Here is a quick introduction to some of the core concepts of Kafka architecture:
+1. Kafka is run as a cluster on one or more servers
+2. Kafka stores streams of records in categories called `topics`. Each record consists of a key, value and a timestamp
+3. Kafka works on the publish-subscribe pattern. Thus, it allows some of the applications to act as `producers` and publish the records to Kafka topics. Similarly, it allows some of the applications to act as `consumers` and subscribe to Kafka topics and process the records produced by it
+4. Alongside, `Producer API` and `Consumer API`, Kafka also offers `Streams API` for an application to work as a stream processor and `Connector API` through which we can connect Kafka to other existing applications and data systems
+
+## Creating Kafka Topics
+
+Let us start by creating a `sample` kafka topic with a single partition and replica. This can be done using the following command:
```
Comment (sidhanthp) [inline posts/hello-world-in-kafka-using-python.md:113]: Explain what it means to have a _single partition and replica_?

File: posts/hello-world-in-kafka-using-python.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+```
+from kafka import KafkaProducer
+producer = KafkaProducer(bootstrap_servers='localhost:9092')
+producer.send('sample', b'Hello, World!')
+producer.send('sample', key=b'message-two', value=b'This is Kafka-Python')
+```
+
+You can now revisit the consumer shell to check if it has received the records sent from the producer through our kafka setup. 
+
+Thus, a simple `Hello, World! in Kafka using Python`.
```
Comment (sidhanthp) [inline posts/hello-world-in-kafka-using-python.md:171]: Are there any drawbacks that can come from using Kafka? What are they?

File: posts/hello-world-in-kafka-using-python.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+> That sounds convincing! But how does Kafka solve the above mentioned challenges and why would one need a system like this?
+
+Kafka is a distributed platform and built for scale, which means it can handle sky-high frequency reads and writes & store huge volumes of data. It ensures that the data is always reliable. It also supports strong mechanisms for recovery from failures. Here are some of key aspects on why one should be using Kafka:
+
+### 1. Simplify the backend architecture
+
+Look at how a complex architecture can be simplified and streamlined with the help of Kafka
+
+![Simple Architecture](../images/kafka-image-three.jpg)
```
Comment (sidhanthp) [inline posts/hello-world-in-kafka-using-python.md:61]: I think putting arrows in both directions from all the items would show that any of the items can be a producer/consumer

File: posts/hello-world-in-kafka-using-python.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+2. If you go ahead and store high frequency data in database systems like SQL or Mongo, it would be hard to introduce and reconstruct a new system or a database on all of the historical data. You lose the flexibility to extend the capabilities of your system by introducing new techologies.
+3. What if you have data processing systems in place to process these events to gain deeper insights? Since these systems wouldn't be capable of handling high frequency reads and you wouldn't have access to the true source of data it is practically impossible to experiment with various data processing, machine learning algorithms on all of the data.
+3. Each application can follow their own data formats, which means that you will need systems for data transformations when there is exchange of data across these applications.
+
+All these problems can be better addressed by bringing a streaming platform like Kafka into the picture. A streaming platform is a system that can perform the following:
+
+1. Store huge amount of data that can be persistent, checksummed and replicated for fault tolerance
```
Comment (sidhanthp) [inline posts/hello-world-in-kafka-using-python.md:43]: Explain why we want our data to be _persistent, checksummed, and replicated_

File: posts/hello-world-in-kafka-using-python.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+![Simple Architecture](../images/kafka-image-two.jpg)
+
+This might look simple, but you're not finished. There are multiple challenges that can arise. 
+
+1. Events like clicks, requests, impressions and searches results in high frequency interaction/requests or data flow to your webserver and your primary database may not be equipped to scale seamlessly. This could introduce high latencies as more and more events pour in to the server.
+2. If you go ahead and store high frequency data in database systems like SQL or Mongo, it would be hard to introduce and reconstruct a new system or a database on all of the historical data. You lose the flexibility to extend the capabilities of your system by introducing new techologies.
+3. What if you have data processing systems in place to process these events to gain deeper insights? Since these systems wouldn't be capable of handling high frequency reads and you wouldn't have access to the true source of data it is practically impossible to experiment with various data processing, machine learning algorithms on all of the data.
```
Comment (sidhanthp) [inline posts/hello-world-in-kafka-using-python.md:38]: Explain what it means to be a _true source of data_? I saw later in the post you wrote that we 'have all the data from all different systems residing at a single place making Kafka a true source of data', but why can't SQL DB's that link to each other also be a true source of data?

**Review by sidhanthp — CHANGES_REQUESTED** [review]
Thanks for the great writeup! Made some grammar changes (already committed to master) and left some feedback for some things to add. If you could redo the charts in LucidChart so they're ready to be uploaded to the site, I'd really appreciate that.

**Comment by zsherman** [comment]
@sidhanthp can you handle the charts? 

@hackintoshrao @abraarsyed I think we're almost ready, could we just address some of Sid's comments?

**Comment by abraarsyed** [comment]
@zsherman Hey there, you are right! We will look into the suggested changes and make the necessary changes as soon as possible :D

**Comment by sidhanthp** [comment]
@abraarsyed Thank you! The charts are finished, just make the changes and submit when you can. Looks like the article is almost done, good job guys!

**Comment by sidhanthp** [comment]
Hey guys, I ended up just making the changes and merging it in. Thank you for helping us with the writeup, it was very well done! You should see it posted on Tuesday morning. We'll give you all the credit in the article and put your twitter handle when we blast it out.

**Comment by hackintoshrao** [comment]
Thank you @sidhanthp . We are working on couple more blogs.
